http_server adapter

mnibecker opened the pull request:
Take `-listen` functionality described in #111 and move it all to its own adapter called `http_server`.

Supports the following options:
- `port` - Not required, defaults to `8080`. The port number the server will listen to
- `method` - Not required, defaults to `POST`. Defines the accepted http method for the created endpoint. Currently only supports `PUT` or `POST`.
- `timeField` - Not required, defaults to `time`. Defines the field to treat as time.

Implementation detail: Points do not actually get emitted until client stops sending server data. This is because I wanted to return a 400 error message if the request is not perfect. I'd be curious to hear any objections over this.

Resolves #111

mnibecker commented:
@rlgomes @demmer

mnibecker commented:
just tested this from `cli`, it works and its totally awesome!

rlgomes commented:
1. lets stick to just single quotes so the soon to be single quote enforcement has to touch less things.
2. I think it would be a nice feature to be able to specify the HTTP method and path to accept requests on and also the ability to dictate which address to bind on. This could simply be handled by acccepting a string to `-listen "http://172.22.23.3:8000/some/special/path" -method DELETE` because who knows if someone wants to record all delete requests they're sending to some other product by splitting the HTTP traffic and hitting an HTTP listener on our end ? (I'm sure there are other good uses for this).

For 2. we could simply create other issues to fill that kind of functionality in I just wanted to bring it up to see if others find that to be a nice thing to have or not.

mnibecker commented:
@rlgomes, agree with the path functionality I'll add that. As for supporting different methods, I'm kind of scratching my head at that. How would `GET` work? `GET` doesn't have a body, you just insert the query params as a point? How would `PUT`/`PATCH` work? `PUT`/`PATCH` are supposed to modify currently existing properties, I'm not quite sure the use here. Same goes for `DELETE`.

rlgomes commented:
@mnibecker I know that I've seen different people doing things that aren't really the "correct" way to use HTTP which does include doing GETS with a body and using PUT instead of  POST as you're directly correctly here.

mnibecker commented:
I'm going to do this with `POST` first, because its what I want. I think we can follow up in a separate pr to add other methods. @go-oleg made some good comments about supporting `PUT` as well. I personally would find it confusing if it did a `GET` request and the server was expecting content in the body as opposed to query params.

mnibecker commented:
Alright moved everything to its own adapter called `http_server`. Updated pr to reflect this.

@rlgomes- I added support for both `POST` and `PUT`. As for supporting other methods I think we can attack this on a separate pr. Same for support path (we discussed this here and couldn't really think of a good use case for path- you can't create multiple listeners on the same port).

rlgomes commented:
+1 looks good just missing docs like the other adapters under: 

https://github.com/juttle/juttle/tree/master/docs/adapters

mnibecker commented:
I will do this and also update the changelog

davidvgalbraith commented on lib/adapters/http_server.js:
```diff
@@ -0,0 +1,133 @@
+
+    start: function() {
+        var self = this;
+        var format;
```
With `format` scoped here, it will be preserved across requests, so if you make a request with one content type and then the next request's content type throws an error on line 56 `format` will be the old content type instead of `undefined`.

mnibecker replied:
yep, great point

davidvgalbraith commented on lib/adapters/http_server.js:
```diff
@@ -0,0 +1,133 @@
+            catch(err) {
+                // type is undefined
+            }
+            try {
```
This should be wrapped in `Promise.try` so you don't have to have separate `catch`es for sync/async errors.

mnibecker replied:
Did not know that's how `Promise.try` worked. Hot tip.

davidvgalbraith commented on lib/adapters/http_server.js:
```diff
@@ -0,0 +1,133 @@
+
+                var emit_points = [];
+
+                var parser =  parsers.getParser(format , {rootPath: self.rootPath});
```
`rootPath` isn't an allowed option?

mnibecker replied:
I suppose I should add `rootPath`

davidvgalbraith commented on lib/adapters/http_server.js:
```diff
@@ -0,0 +1,133 @@
+                    }
+
+                    self.parseTime(points, self.timeField);
+                    emit_points = emit_points.concat(points);
```
Why buffer all the points like this instead of emitting them here?

mnibecker replied:
Because if there's an error parsing a line I wanted to throw back a `400` and reject the entire request. The alternative was returning a `200` and not ingesting some of the points, which felt weird to me.

davidvgalbraith commented on test/adapters/http_server.spec.js:
```diff
@@ -0,0 +1,298 @@
+        })
+        .then(function(prog) {
+            program = prog;
+            programFinish = run_juttle(prog);
```
This is a neat testing strategy!

mnibecker commented:
Great comments @davidvgalbraith. I'll throw up a pr that addresses your comments/fixes
